test/property.cpp

```diff
@@ -15,8 +15,6 @@
 BOOST_AUTO_TEST_SUITE(test_property)
 
 BOOST_AUTO_TEST_CASE( content_type_ref ) {
-    static_assert(std::is_convertible<mqtt::v5::property::content_type, mqtt::v5::property::content_type_ref>::value, "");
-    static_assert(std::is_convertible<mqtt::v5::property::content_type_ref, mqtt::v5::property::content_type>::value, "");
```

Comment:
I agree with adding explicit to the constructor. 
But it avoids the following code on #279 you mentioned:

```cpp
[](..., std::vector<mqtt::v5::property_variant> const& props)
{
    std::optional<mqtt::v5::property::correlation_data> corData;
    std::optional<mqtt::v5::property::response_topic> respTopic;
    for(mqtt::v5::property_variant prop : props)
    {
        std::visit([&](auto&& arg)
        {
            using T = std::decay_t<decltype(arg)>;
            if constexpr (std::is_convertible_v<T, mqtt::v5::property::correlation_data>)
            {
                corData = arg;
            }
            else if constexpr (std::is_convertible_v<T, mqtt::v5::property::response_topic>)
            {
                respTopic = arg;
            }
        }, prop);
    }
}
```

But I think that you can use `std::is_constructible` instead of `std::is_convertible`. See https://wandbox.org/permlink/oWSHhWlGYxwn7Ufd

So could you replace as follows instead of removing them.

```cpp
    static_assert(std::is_constructible<mqtt::v5::property::content_type, mqtt::v5::property::content_type_ref>::value, "");	
    static_assert(std::is_constructible<mqtt::v5::property::content_type_ref, mqtt::v5::property::content_type>::value, "");
```

Reply:
I looked over the comment

> -- Move several static assertions into the code itself, instead of unit tests, to ensure the assertions are always compiled.

Could you move somewhere (maybe property.hpp) to static_assert (is_constructible).

Reply:
I looked over the comment

> -- Move several static assertions into the code itself, instead of unit tests, to ensure the assertions are always compiled.

Could you move somewhere (maybe property.hpp) to static_assert (is_constructible).

Reply:
Done

Reply:
I kept the "is_convertible_v" and also added "is_constructible_v".

Refs should be implicitly constructible from non-refs.

Both versions should be explicitly constructible from the other.

Reply:
Ok, that means you only add `explicit` to **non_ref** class constructor that has `ref` parameter.

Reply:
Right. Because constructing a _ref type does not cause an allocation, so it can be entirely optimized away by the compiler (probably).

Where as constructing a non_ref type always allocates, so the user should only have that happen if they really intended it to.